tests/test_command_generation.py

```diff
@@ -168,30 +168,124 @@ class FakeLinter(Linter):
             cmd = ('fake_linter_1',)
             defaults = {'selector': None}
 
-        settings = {'executable': 'my_linter'}
-        result = ['my_linter']
+        EXECUTABLE = 'my_linter'
+        RESULT = ['resolved']
+        settings = {'executable': EXECUTABLE}
 
         linter = FakeLinter(self.view, settings)
-        # XXX: We probably don't need to test `can_exec`
-        # - Popen will also throw and show the error panel
-        # - User could just set e.g. 'linter.exe', and the OS will use PATH
-        #   to resolve that automatically
-        # - We don't check for arrays, see below
-        with when(util).can_exec('my_linter').thenReturn(True), \
-             expect(linter)._communicate(result, ...):  # noqa: E127
+        when(util).which('my_linter').thenReturn('resolved')
+        with expect(linter)._communicate(RESULT, ...):
             linter.lint(INPUT, VIEW_UNCHANGED)
 
     def test_executable_is_set_to_an_array(self):
         class FakeLinter(Linter):
             cmd = ('fake_linter_1',)
             defaults = {'selector': None}
 
-        settings = {'executable': ['my_interpreter', 'my_linter']}
-        result = ['my_interpreter', 'my_linter']
+        EXECUTABLE = ['my_interpreter', 'my_linter']
+        RESULT = ['resolved', 'my_linter']
+        settings = {'executable': EXECUTABLE}
 
         linter = FakeLinter(self.view, settings)
-        with expect(linter)._communicate(result, ...):
+        when(util).which('my_interpreter').thenReturn('resolved')
+        with expect(linter)._communicate(RESULT, ...):
+            linter.lint(INPUT, VIEW_UNCHANGED)
+
+    def test_unhappy_given_short_name(self):
+        class FakeLinter(Linter):
+            cmd = ('fake_linter_1',)
+            defaults = {'selector': None}
+
+        EXECUTABLE = 'my_linter'
+        settings = {'executable': EXECUTABLE}
+
+        linter = FakeLinter(self.view, settings)
+        when(linter).notify_failure().thenReturn(None)
+        when(util).which(...).thenReturn(None)
+        when(linter_module.logger).error(...)
+
+        try:
             linter.lint(INPUT, VIEW_UNCHANGED)
+        except linter_module.PermanentError:
+            pass
+
+        verify(linter_module.logger).error(
+            "You set 'executable' to 'my_linter'.  "
+            "However, 'which my_linter' returned nothing."
```

Comment:
> "However, 'which my_linter' returned nothing."

This could be better. Can you give me a hint. Like "However, 'which my_linter' returned nothing.\nTry setting an absolute path to the binary. Also see troubleshooting guide."

Reply:
👍 A good error message also proposes a possible solution and it's sorely missing here.